### PR TITLE
[CALCITE-2450] Normalize RexCall predicates when computing digest

### PR DESCRIPTION
**Pull request by vlsi**
`$0=$1` and `$1=$0` have the same semantics, so it is worth normalizing them to reduce the planning time.

See https://issues.apache.org/jira/browse/CALCITE-2450

### CODE REVIEW

File: core/src/main/java/org/apache/calcite/rex/RexCall.java [hunk 1]
```diff
@@ -127,7 +143,23 @@ protected final StringBuilder appendOperands(StringBuilder sb) {
           includeType = RexDigestIncludeType.NO_TYPE;
         }
       }
-      sb.append(((RexLiteral) operand).computeDigest(includeType));
+      operandDigests.add(((RexLiteral) operand).computeDigest(includeType));
+    }
+    int totalLength = (operandDigests.size() - 1) * 2; // commas
+    for (String s : operandDigests) {
```
Comment (danny0405) [inline core/src/main/java/org/apache/calcite/rex/RexCall.java]: How about the operands are empty ?

**Comment by danny0405** [comment]
Please **do not** merge the PR before we solve the plan compatibility.

File: core/src/main/java/org/apache/calcite/rex/RexCall.java (hunk 1 above)
Comment (vlsi) [reply core/src/main/java/org/apache/calcite/rex/RexCall.java]: They must not be empty, because `AND()`, `=()` and so on means nothing.

**Comment by vlsi** [comment]
@danny0405 , what do you mean by "solve the plan compatibility"?

**Comment by danny0405** [comment]
> @danny0405 , what do you mean by "solve the plan compatibility"?

The plan changes, with same semantic but no  promotion. "The plan compatibility" means we should keep as the same plan for same query/relational expression for different Calcite version.

**Comment by vlsi** [comment]
> The plan changes, with same semantic but no promotion

The key improvement here is performance improvement.
There are several tests that get faster.
For instance: org.apache.calcite.test.RelMetadataTest#testPullUpPredicatesForExprsItr
There's an assert:

```
assertThat(inputSet.pulledUpPredicates.size(), is(18));
```

After the change it becomes `12`, so the test extracts less predicates, thus the planning becomes faster.


Note: even though the plan representation change, it is "one-time" change.
In other words, the plan would be sorted in exactly the same order, so the change would make plans **more stable** in the future.

> we should keep as the same plan for same query/relational expression for different Calcite version

It can't always be like that. A change to the optimizer might produce a slightly different plan.
This PR improves performance across different tests, so it is a true performance improvement.

However, I do not see the way to fully reproduce "old plan representation" because "old representation" is "a random one".

**Comment by danny0405** [comment]
How many changed plans there can have a performance promotion ? I would say very few. We can do some promotion code and that is awesome, but please **do not bring in** extra burden to the users that do not need this feature.

> In other words, the plan would be sorted in exactly the same order, so the change would make plans more stable in the future.

I don't think so, isn't keep the rex calls the sequence just as what user writes them more natural, more readable more stable ?

**Comment by vlsi** [comment]
> isn't keep the rex calls the sequence just as what user writes them more natural, more readable more stable ?

RexSimplify can easily adjust the nodes, put elements to HashSet, and so on.

> I would say very few

How do you know that?
https://issues.apache.org/jira/browse/CALCITE-2223 is quite popular.

File: core/build.gradle.kts [hunk 2]
```diff
@@ -41,6 +41,7 @@ dependencies {
 
     api("com.fasterxml.jackson.core:jackson-annotations")
     api("org.apache.calcite.avatica:avatica-core")
+    api("org.apiguardian:apiguardian-api")
```
Comment (yanlin-Lynn) [inline core/build.gradle.kts]: this seems to be another feature that is unrelated with this pr

Comment (vlsi) [reply core/build.gradle.kts]: It is related. The dependency is needed to add `@API(since = "1.22",` annotation to the methods that are added in this PR.

Comment (danny0405) [reply core/build.gradle.kts]: Can we write our own annotation ? I would expect to not introduce a dependency just for a annotation, using our own annotations keep better compatibility and reduce conflicts.

File: core/src/main/java/org/apache/calcite/adapter/enumerable/EnumerableHashJoin.java [hunk 3]
```diff
@@ -120,24 +121,23 @@ public static EnumerableHashJoin create(
...
-      return planner.getCostFactory().makeCost(rowCount, 0, 0).multiplyBy(.01d);
+      // For semi- and anti- the right input is hashed
+      rowCount += rightRowCount * 1.1d;
+      return costFactory.makeCost(rowCount, 0, 0).multiplyBy(.01d);
     } else {
-      return planner.getCostFactory().makeCost(rowCount, 0, 0);
+      // Lots of plans assumed the left join was hashed, so
+      // we adjust a cost accordingly
+      rowCount += leftRowCount * 1.1d;
+      return costFactory.makeCost(rowCount, 0, 0);
```
Comment (danny0405) [inline core/src/main/java/org/apache/calcite/adapter/enumerable/EnumerableHashJoin.java]: Why a hash cost is `row_cnt * 1.1` ?

File: core/src/main/java/org/apache/calcite/config/CalciteSystemProperty.java [hunk 4]
```diff
@@ -83,6 +83,10 @@
   public static final CalciteSystemProperty<Boolean> ENABLE_STREAM =
       booleanProperty("calcite.enable.stream", true);
 
+  /** Whether streaming is enabled in the default planner configuration. */
+  public static final CalciteSystemProperty<Boolean> ENABLE_REX_DIGEST_NORMALIZE =
```
Comment (danny0405) [inline core/src/main/java/org/apache/calcite/config/CalciteSystemProperty.java]: Fix the java doc.

File: core/src/main/java/org/apache/calcite/rel/externalize/RelJsonWriter.java [hunk 5]
```diff
@@ -140,6 +142,17 @@ public RelWriter done(RelNode node) {
   public String asString() {
     final Map<String, Object> map = jsonBuilder.map();
     map.put("rels", relList);
-    return jsonBuilder.toJsonString(map);
+    AtomicInteger skipNormalize = null;
+    if (getDetailLevel() != SqlExplainLevel.DIGEST_ATTRIBUTES) {
+      skipNormalize = RexNode.DESCRIBE_WITHOUT_NORMALIZE.get();
+      skipNormalize.incrementAndGet();
+    }
+    try {
```
Comment (danny0405) [inline core/src/main/java/org/apache/calcite/rel/externalize/RelJsonWriter.java]: This wrapper appears in many places, can we write a method to encapsulate that ?

File: core/src/main/java/org/apache/calcite/rex/RexCall.java [hunk 6]
```diff
@@ -73,6 +85,23 @@
     SIMPLE_BINARY_OPS = Sets.immutableEnumSet(kinds);
   }
 
+  private static final Set<SqlKind> SIMPLE_COMMUTATIVE_OPS;
+
+  static {
+    SIMPLE_COMMUTATIVE_OPS = Sets.immutableEnumSet(
+        SqlKind.AND, SqlKind.OR, SqlKind.PLUS, SqlKind.TIMES, SqlKind.EQUALS, SqlKind.NOT_EQUALS,
+        SqlKind.IS_DISTINCT_FROM, SqlKind.IS_NOT_DISTINCT_FROM,
+        SqlKind.GREATEST, SqlKind.LEAST);
```
Comment (danny0405) [inline core/src/main/java/org/apache/calcite/rex/RexCall.java]: Compared to list these `SqlKind` here, i would suggest to new a fresh new `SqlKind` named `Symmetrical` in file `SqlKind.java` which include these symmetrical operators.

File: core/src/main/java/org/apache/calcite/rex/RexCall.java [hunk 7]
```diff
@@ -155,15 +237,15 @@ private static boolean equalSansNullability(RelDataType a, RelDataType b) {
         && x.startsWith(y);
   }
 
-  protected @Nonnull String computeDigest(boolean withType) {
+  protected @Nonnull String computeDigest(boolean withType, boolean withNormalize) {
     final StringBuilder sb = new StringBuilder(op.getName());
```
Comment (danny0405) [inline core/src/main/java/org/apache/calcite/rex/RexCall.java]: I would suggest to introduce a new config class for computing digest, something like `computeDigest(RexCall.Config)` to give the interface better extensibility.

Comment (vlsi) [reply core/src/main/java/org/apache/calcite/rex/RexCall.java]: Frankly speaking, I would refrain from extracting configuration classes and making this configurable unless there are clear use cases

File: core/src/main/java/org/apache/calcite/adapter/enumerable/EnumerableHashJoin.java (hunk 3 above)
Comment (vlsi) [reply core/src/main/java/org/apache/calcite/adapter/enumerable/EnumerableHashJoin.java]: The idea was hashtable creation has a linear cost, but it is a bit more than just processing the entries.

In practice, this needs adjustments, as org.apache.calcite.test.JdbcTest#testAlmostBushy plan must be flipped, but currently, it does not flip, so hash join should be adjusted somehow.

Probably I'm missing hash lookup cost.

File: core/build.gradle.kts (hunk 2 above)
Comment (vlsi) [reply core/build.gradle.kts]: Do you expect the same annotation to be used in Avatica and Calcite?
Does that mean the annotation should be declared in Avatica, and can be used in Calcite only after Avatica is released? It does sound too complicated.

On the other hand, https://github.com/apiguardian-team/apiguardian is a more-or-less clear dependency that has already been tested (e.g. it is used in JUnit5).

In other words, `apiguardian-api` is already available (and it can be easily used by other projects for consistent API declaration!) while Calcite's annotation is nowhere near.

File: core/src/main/java/org/apache/calcite/adapter/enumerable/EnumerableHashJoin.java (hunk 3 above)
Comment (vlsi) [reply core/src/main/java/org/apache/calcite/adapter/enumerable/EnumerableHashJoin.java]: The problem here is not 1.1, but the problem is join cardinality estimation.
Calcite tends to over-estimate the join cardinality, thus leftRowCount and rightRowCount tend to be noise when compared with the estimated join cardinality, thus all those `+= leftRowCount * 1.1d` change nothing.

So a series of patches (https://issues.apache.org/jira/browse/CALCITE-3666, https://issues.apache.org/jira/browse/CALCITE-3661) is required before this costing could provide any meaningful results :(

File: core/src/main/java/org/apache/calcite/config/CalciteSystemProperty.java (hunk 4 above)
Comment (vlsi) [reply core/src/main/java/org/apache/calcite/config/CalciteSystemProperty.java]: fixed, thanks

File: core/src/main/java/org/apache/calcite/rel/externalize/RelJsonWriter.java (hunk 5 above)
Comment (vlsi) [reply core/src/main/java/org/apache/calcite/rel/externalize/RelJsonWriter.java]: I would consider that to be on the hot path (especially in RexCall), so I would refrain from allocating extra objects there.

Adding a method like `withoutNormalize(Runnable)` won't make a huge readability difference, yet it would add overhead. The API is not intended for public use, so I'm inclined to leave it as is.

File: core/src/main/java/org/apache/calcite/rex/RexCall.java (hunk 7 above)
Comment (vlsi) [reply core/src/main/java/org/apache/calcite/rex/RexCall.java]: I've reverted `, boolean withNormalize` argument, and replaced it with `protected static boolean needNormalize()`.

File: core/src/main/java/org/apache/calcite/rel/externalize/RelJsonWriter.java (hunk 5 above)
Comment (vlsi) [reply core/src/main/java/org/apache/calcite/rel/externalize/RelJsonWriter.java]: I've extracted that to `class RexNode { ... public static Closeable skipNormalize()` which is to be used in try-with-resources.

File: core/src/main/java/org/apache/calcite/rex/RexCall.java (hunk 6 above)
Comment (vlsi) [reply core/src/main/java/org/apache/calcite/rex/RexCall.java]: Moved to `SqlKind#SYMMETRICAL` and `SqlKind#SYMMETRICAL_SAME_ARG_TYPE` (which is for cases like `PLUS`, and `TIMES` which can be re-arranged only in case the argument types are the same).
In other words, timestamp + interval vs interval + timestamp are different expressions.

**Comment by vlsi** [comment]
@danny0405 , I don't understand what do you mean by `needs-a-final-review` label.
The change has been tested both in isolation and with the cost changes, and it behaves well.

I've added the tests and added `try-with-resources` API which simplifies the use, and `ThreadLocal` is now `private`.

**Comment by danny0405** [comment]
> @danny0405 , I don't understand what do you mean by `needs-a-final-review` label.
> The change has been tested both in isolation and with the cost changes, and it behaves well.
> 
> I've added the tests and added `try-with-resources` API which simplifies the use, and `ThreadLocal` is now `private`.
> 
> So I'm inclined to merge the PR.

The PR looks good from my side, go ahead ~

File: core/src/main/java/org/apache/calcite/rex/RexCall.java [hunk 8]
```diff
@@ -127,11 +137,83 @@ protected final StringBuilder appendOperands(StringBuilder sb) {
...
+      final RelDataType firstType = operands.get(0).getType();
+      for (int i = 1; i < operands.size(); i++) {
+        if (!equalSansNullability(firstType, operands.get(i).getType())) {
+          // Arguments have different type, thus they must not be sorted
+          return;
+        }
+      }
+      // fall through: order arguments below
+    } else if (!SqlKind.SYMMETRICAL.contains(kind)
+        && (kind == kind.reverse()
```
Comment (danny0405) [inline core/src/main/java/org/apache/calcite/rex/RexCall.java]: I have left the comments in another PR https://github.com/apache/calcite/pull/1740

**Comment by danny0405** [comment]
Overall looks pretty good, i have left some minor comments in PR https://github.com/apache/calcite/pull/1740
